### PR TITLE
App Life Cycle Hooks PROMOTE getState function RATHER-THAN appState object

### PR DESCRIPTION
**Pull request by sylvainlg**
In some cases, the instantaneous `appState` is not enough.
In my sample, we imagine a axios configuration trough interceptors.

When the user is not authenticated, the request interceptor is not modifed.
When the user is authenticated, the jwt token is set.

Let's see how to do it in a feature context :

(It works the same way with appInit)

**configureAxios.js**
```js
export default (fassets, appState) => {
  // We intercept Axios calls to add authentication headers
  axios.interceptors.request.use(
    async (requestConfig) => {

      if (fassets.auth.isAuth(appState)) {

        const authHeaders = {
          apiKey: Config.API_KEY,
          Authentication: `Bearer ${fassets.auth.getJwtToken(appState)}`,
        }

        const { headers } = requestConfig
        const newHeaders = {
          ...headers,
          ...authHeaders,
        }
        requestConfig.headers = newHeaders

        return requestConfig

      } else {
        return requestConfig

      }
    },
    (error) => {
      console.error(error)
      return error
    },
  )
}
```

**kernel/feature.js**
```js
export default createFeature({
  name: 'kernel',
  appDidStart: ({fassets, appState}) => { configureAxios(fassets, appState) }
```

In that situation, appState *always* have the same value : the value he had when appDidStart was call. If the user authenticates, the interceptor doesn't know about it and the app ins't working.

------
Now, if we provide the `appDidStart` the `getState` instead of the `appState`, the `configureAxios` is able to retrieve the up-to-date state and works as expected during the all app execution.

In that case, the code will be : 

**configureAxios.js**
```js
export default (fassets, getState) => {
  // We intercept Axios calls to add authentication headers
  axios.interceptors.request.use(
    async (requestConfig) => {

      if (fassets.auth.isAuth(getState())) {

        const authHeaders = {
          apiKey: Config.API_KEY,
          Authentication: `Bearer ${fassets.auth.getJwtToken(getState())}`,
        }

        const { headers } = requestConfig
        const newHeaders = {
          ...headers,
          ...authHeaders,
        }
        requestConfig.headers = newHeaders

        return requestConfig

      } else {
        return requestConfig

      }
    },
    (error) => {
      console.error(error)
      return error
    },
  )
}
```

**kernel/feature.js**
```js
export default createFeature({
  name: 'kernel',
  appDidStart: ({fassets, getState}) => { configureAxios(fassets, getState) }
```

### CODE REVIEW

**Comment by KevinAst** [comment]
My assessment of this change is that it appears to be a legitimate enhancement!  Thanks @sylvainlg for this catch.

If I were to summarize this, it would be:

<ul>

For **feature-u** redux users, the Application Life Cycle hooks that utilize the redux `appState` parameter are now being passed the redux `getState` function.  To obtain the `appState` you merely invoke: `getState()`.  This gives an async or long-running process access to the most current state changes _(over time)_.

</ul>

I will plan on including this change in my next production release.


**Sidebar**:

<ul>

With **no** knowledge of your app, it appears to me that your `configureAxios` is using the `appDidStart` hook solely to access the redux state.  Is this correct?

I suppose that is OK.  The bottom line is you need access to the redux state outside of a conventional react component.  I suppose that the access provided by **feature-u**'s "Application Life Cycle hooks" is as good as any :-)

At minimum, it seems that it may be better to use the `appInit` hook, from the perspective introducing it earlier in your process.  It seems like you would want the logic that manages the axios headers configured as early as possible.  With that being said, it may be a moot point, since the logic no-ops when the user has not yet authenticated.

Thoughts?

</ul>

**Minor Point**:

<ul>

I would recommend using named parameters in your `configureAxios` function, simply because it would have the same signature to the life-cycle hook.  With that done:

**this**:
```js
  appDidStart: ({fassets, getState}) => { configureAxios(fassets, getState) }
```
**would become this**:
```js
  appDidStart: configureAxios
```

</ul>


**Release Schedule Note**:

<ul>

Please be aware that due to my current work load on other projects and overall schedule, this release will be delayed just a bit.  I will need to update the versioned docs, as well as provide some additional unit tests.  I may decide to incorporate some other backlogged features in this upcoming release.

In the mean time, I can think of two work-arounds:

- either use your patched **feature-u** version,

- or temporarily expose your `reduxAspect` _(say through a named export)_, and access the `getState` function as follows:

  `const getState = reduxAspect.getReduxStore().getState();`

</ul>

**Comment by KevinAst** [comment]
Hello @sylvainlg,

1. I have fully integrated/documented/tested this PR in my [next7 branch](https://github.com/KevinAst/feature-u/blob/next7/docs/history.md#v3_0_0).

   **FYI**: I took this opportunity to remove the coupling between **feature-u** and **feature-redux**, by introducing a new `Aspect.injectParamsInHooks()` method, which allows any Aspect Plugin to promote their own `namedParams` into **feature-u**'s Application Life Cycle Hooks.

   **Bottom Line**: Once I release this, it will require an upgrade to BOTH** **feature-u** and **feature-redux**.

2. **Release Time Frame**:

   My best guess for release will be "mid to late" next week.

   I am working on a number of additional enhancements _(both **feature-u** and plugins)_, and am wanting to coordinate their release together :-)

**Comment by sylvainlg** [comment]
Hello,

Sorry for my late response.

You have right about the `appInit`, I guess I tried each hooks in order to get the correct app state and stayed with the wrong one.

The `appDidStart: configureAxios` looks better indeed !

Nice `Aspect.injectParamsInHooks()`, we may don't require redux for long because we are going on [GraphQL with Apollo Client](https://www.apollographql.com/).
With that change, we don't have to handle the unused redux integration.

I look at the release note, if you want to avoid a breaking change, it's possible to pass either `appState` and `getState` to hooks. But it bring more complexity to the API.

**Comment by KevinAst** [comment]
Hello @sylvainlg,

1. **feature-u** V3 has been released which includes this PR.

   This PR was fully integrated/documented/tested.

   **FYI**: I took this opportunity to remove the coupling between **feature-u** and **feature-redux**, by introducing a new `Aspect.injectParamsInHooks()` method, which allows any Aspect Plugin to promote their own `namedParams` into **feature-u**'s Application Life Cycle Hooks.

   **Don't Forget** The **feature-u** V3 release also requires upgrading **feature-redux** to V3.

2. regarding [**feature-redux** PR 16](https://github.com/KevinAst/feature-redux/pull/16)

   **feature-redux** V3 now accepts client-supplied enhancers!

3. regarding [**feature-redux** Issue 15](https://github.com/KevinAst/feature-redux/issues/15)

   - Per the discussion points found there, **feature-u** V3 has relaxed the Aspect Plugin required parameters.  Please refer to the **"No Single Aspect Method is Required"** discussion in the [Aspect Life Cycle Methods](https://feature-u.js.org/cur/extending.html#aspect-life-cycle-methods).

     As a result, your [**feature-redux-persist**](https://github.com/sylvainlg/feature-redux-persist) plugin can omit the dummy `validateFeatureContent` and `assembleFeatureContent` methods.  You only need to specify the `name` for identity purposes.
     
     **You may want to consider changing the internals of your plugin to reflect this.**

     **P.S.** Your **feature-redux-persist** plugin is mentioned in the **feature-u** docs _(explaining these concepts)_ ... it's in the **"No Single Aspect Method is Required"** discussion _(mentioned above)_ ... _congratulations_!! :-)

   - Also, within that discussion:

     The **feature-u** docs now correctly reflect Aspect Plugin configuration accomplished through construction parameters rather than the singleton multi-step process.  Please refer to: [Custom Aspect Plugins](https://feature-u.js.org/cur/extending.html#custom-aspect-plugins) and [`Aspect.genesis()`](https://feature-u.js.org/cur/extending.html#aspectgenesis)

     This is what your new [**feature-redux-persist**](https://github.com/sylvainlg/feature-redux-persist) plugin is doing!

     I have also retrofitted my [**feature-router**](https://github.com/KevinAst/feature-router) to this new philosophy. Take a peek when you have a chance.

4. **P.S.** Can I get your email address?  I want to send a handful of **feature-u** users a request to review an enhancement I am contemplating.  Either reply to this PR, or send me your email to: kevin@wiiBridges.com
